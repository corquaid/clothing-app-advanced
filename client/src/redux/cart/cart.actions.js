import CartActionTypes from "./cart.types";

export const toggleCartHidden = () => ({
    type: CartActionTypes.TOGGLE_CART_HIDDEN, // no need to pass an optional payload as there is no new state data, the state is only being toggled true / false
});

export const addItem = item => ({
    // item is passed into addItem function as designated as the payload for the cart reducer
    type: CartActionTypes.ADD_ITEM,
    payload: item,
});

export const clearItemFromCart = item => ({
    type: CartActionTypes.CLEAR_ITEM_FROM_CART,
    payload: item,
});

export const removeItem = item => ({
    type: CartActionTypes.REMOVE_ITEM,
    payload: item,
});

export const clearCart = () => ({
    type: CartActionTypes.CLEAR_CART,
});
