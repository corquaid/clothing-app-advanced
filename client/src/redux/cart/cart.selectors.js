import { createSelector } from 'reselect';

// Input selector
const selectCart = state => state.cart; // this selects only the cart component of state

// Replacing mapstatetoprops in cart-dropdown
export const selectCartItems = createSelector( // 1st arg is the input selectors, 2nd arg is a function
  [selectCart],
  cart => cart.cartItems
);

export const selectCartHidden = createSelector(
  [selectCart],
  cart => cart.hidden
)

// Replacing mapstatetoprops with reduce function in cart-icon
export const selectCartItemsCount = createSelector(
  [selectCartItems],
  cartItems => cartItems.reduce((totalQuantity, cartItem) => totalQuantity + cartItem.quantity , 0)
);

// Another reduce function to calculate the total cost of items in the cart
export const selectCartTotal = createSelector(
  [selectCartItems],
  cartItems => cartItems.reduce((totalPrice, cartItem) => totalPrice + (cartItem.price * cartItem.quantity), 0)
);