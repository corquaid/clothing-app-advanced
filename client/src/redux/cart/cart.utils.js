// Function to check if item to be added to the cart already exists in the cart or not

export const addItemToCart = (cartItems, cartItemToAdd) => {
  const existingCartItem = cartItems.find( //.find() check will return the first matching value based on the condition passed in
    cartItem => cartItem.id === cartItemToAdd.id // check if IDs match
    );

    if (existingCartItem) { // this if block will not run at all if the item is new, will fgo directly to return statement below
      return cartItems.map(cartItem => // needs to be a map function to return a brand-new array to force React to re-render
        cartItem.id === cartItemToAdd.id // same check as above
        ? { ...cartItem, quantity: cartItem.quantity + 1 } // return new object including the item plus the incremented quantity value
        : cartItem // return cartItem as is with no modification
        )
    }
    return [...cartItems, {...cartItemToAdd, quantity: 1 }]; // if item to be added is not found above, then return a new array with all existing items plus the new item with an initial quantity of 1
};

// Funtion to remove/decrement items from cart using checkout page arrows
export const removeItemFromCart = (cartItems, cartItemToRemove) => {
  const existingCartItem = cartItems.find(
    cartItem => cartItem.id === cartItemToRemove.id
  );

  if (existingCartItem.quantity === 1) { // if there is only one of these items in the cart
    return cartItems.filter(cartItem => cartItem.id !== cartItemToRemove.id) // keep the items whose id does not match the id to remove
  }

  return cartItems.map( // if there is more than one of the item to be removed in the cart
    cartItem => cartItem.id === cartItemToRemove.id ?
    {...cartItem, quantity: cartItem.quantity - 1}
    : cartItem
  );
};