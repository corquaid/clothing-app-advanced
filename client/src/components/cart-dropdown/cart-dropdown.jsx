import React from 'react';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { withRouter } from 'react-router-dom';

import CustomButton from '../custom-button/custom-button';
import CartItem from '../cart-item/cart-item';
import CheckoutPage from '../../pages/checkout/checkout';
import { selectCartItems } from '../../redux/cart/cart.selectors';
import { toggleCartHidden } from '../../redux/cart/cart.actions';

import './cart-dropdown.scss';

const CartDropdown = ({ cartItems, history, dispatch }) => ( // cartItems from props, history also, dispatch refers to mapDispatchToProps, it can be passed in singly from connect() 
  <div className='cart-dropdown'>
    <div className='cart-items'>
      {
        cartItems.length ? // if cartItems.length is empty, this is zero or FALSY
        cartItems.map(cartItem => <CartItem key={cartItem.id} item={cartItem} />) // map through cartItems and create a div for each cartItem
        : <span className='empty-message'>You don't have anything in your cart</span>
      }
    </div>
    <CustomButton onClick={() => {
      history.push('/checkout') // routing to the checkout page
      dispatch(toggleCartHidden()) // toggle function to hide the cart dropdown when the button is clicked
      }}>
        GO TO CHECKOUT
      </CustomButton>
    
    {/* this history one is a bit odd */}
  </div>
);

const mapStateToProps = createStructuredSelector({ 
  cartItems: selectCartItems // from selector file
})

export default withRouter(connect(mapStateToProps)(CartDropdown)); // connect () automatically passes a mapdispatchtoprops into the component if a 2nd argument is not specified after mapStateToProps meaning that the dispatch properties are accessible inside the component