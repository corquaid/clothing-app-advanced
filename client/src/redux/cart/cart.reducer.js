import CartActionTypes from './cart.types';

import { addItemToCart, removeItemFromCart } from './cart.utils';

const INITIAL_STATE = {
  hidden: true, // dropdown is hidden when website loads initially
  cartItems: [] // cart is initially an empty array
}

const cartReducer = (state = INITIAL_STATE, action) => {
  switch (action.type) {
    case CartActionTypes.TOGGLE_CART_HIDDEN:
      return {
        ...state,
        hidden: !state.hidden // bang ! operator converts the hidden state to the opposite of the current true/false value
      };
    case CartActionTypes.ADD_ITEM:
      return {
        ...state,
        cartItems: addItemToCart(state.cartItems, action.payload) // external utils function to create new cart array with new item to add, taken from the payload value
      };
    case CartActionTypes.CLEAR_ITEM_FROM_CART:
      return {
        ...state,
        cartItems: state.cartItems.filter( // function that removes item from cart, returning a new array with the item to be removed filtered out
          cartItem => cartItem.id !== action.payload.id // keep any items whose id does not match the item that we are trying to remove
        )
      };
    case CartActionTypes.REMOVE_ITEM:
      return {
        ...state,
        cartItems: removeItemFromCart(state.cartItems, action.payload)
      };
    case CartActionTypes.CLEAR_CART:
      return {
        ...state,
        cartItems: []
      }
    default:
      return state;
  }
}

export default cartReducer;