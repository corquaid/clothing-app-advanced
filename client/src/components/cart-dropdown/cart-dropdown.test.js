import React from "react";
import { shallow } from "enzyme";

import { CartDropdown } from "./cart-dropdown";

import { toggleCartHidden } from "../../redux/cart/cart.actions";

// Test group
describe("CartDropdown component", () => {
    // set up variables to mock props
    let wrapper;
    let mockHistory;
    let mockDispatch;

    const mockCartItems = [{ id: 1 }, { id: 2 }, { id: 3 }];

    // Run this entire function to mock the "Go To Checkout" button click event
    beforeEach(() => {
        mockHistory = {
            push: jest.fn(),
        };

        mockDispatch = jest.fn(); // mock toggleCartHidden state change

        // completed mockProps object for the CartDropdown component
        const mockProps = {
            cartItems: mockCartItems,
            history: mockHistory,
            dispatch: mockDispatch,
        };

        wrapper = shallow(<CartDropdown {...mockProps} />);
    });

    // Start defining tests

    it("Should render CartDropdown component", () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('should call history.push when button is clicked', () => {
     
        wrapper.find('[id="button"]').simulate('click');
        expect(mockHistory.push).toHaveBeenCalled();
        expect(mockDispatch).toHaveBeenCalledWith(toggleCartHidden())
    })

    it('Should render an equal number of items as the cartItems prop', () => {
        expect(wrapper.find('[id="cart-item"]').length).toEqual(mockCartItems.length)
    })

    it('Should render empty message container if the cart is empty', () => {
        const mockProps = {
            cartItems: [],
            history: mockHistory,
            dispatch: mockDispatch
        }

        const wrapperTwo = shallow(<CartDropdown {...mockProps} />)
        expect(wrapperTwo.exists('.empty-message')).toBe(true)
    })
});
