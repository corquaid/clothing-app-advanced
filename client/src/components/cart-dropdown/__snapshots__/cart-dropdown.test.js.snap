// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartDropdown component Should render CartDropdown component 1`] = `ShallowWrapper {}`;
